# HG changeset patch
# User Gabriel Luong <gabriel.luong@gmail.com>
# Parent  4a14671935e6bc31b3f7c509190bbd30d25de4d2
Adds a new pref for the new markup view.

diff --git a/devtools/client/inspector/inspector.js b/devtools/client/inspector/inspector.js
--- a/devtools/client/inspector/inspector.js
+++ b/devtools/client/inspector/inspector.js
@@ -732,16 +732,42 @@ Inspector.prototype = {
             this.layoutview = new LayoutView(this, this.panelWin);
           }
 
           return this.layoutview.provider;
         }
       },
       defaultTab == layoutId);
 
+    if (Services.prefs.getBoolPref("devtools.markup.new-frontend-enabled")) {
+      // Inject a lazy loaded react tab by exposing a fake React object
+      // with a lazy defined Tab thanks to `panel` being a function
+      let markupId = "markupview";
+      let markupTitle = INSPECTOR_L10N.getStr("inspector.panelLabel.markupView");
+      this.sidebar.addTab(
+        markupId,
+        markupTitle,
+        {
+          props: {
+            id: markupId,
+            title: markupTitle
+          },
+          panel: () => {
+            if (!this.markupview) {
+              const MarkupView =
+                this.browserRequire("devtools/client/inspector/markup/new-markup");
+              this.markupview = new MarkupView(this, this.panelWin);
+            }
+
+            return this.markupview.provider;
+          }
+        },
+        defaultTab == markupId);
+    }
+
     if (Services.prefs.getBoolPref("devtools.changesview.enabled")) {
       // Inject a lazy loaded react tab by exposing a fake React object
       // with a lazy defined Tab thanks to `panel` being a function
       let changesId = "changesview";
       let changesTitle = INSPECTOR_L10N.getStr("inspector.sidebar.changesViewTitle");
       this.sidebar.addTab(
         changesId,
         changesTitle,
@@ -1267,16 +1293,20 @@ Inspector.prototype = {
     this.target.off("thread-resumed", this.updateDebuggerPausedWarning);
     this._toolbox.off("select", this.updateDebuggerPausedWarning);
 
     for (let [, panel] of this._panels) {
       panel.destroy();
     }
     this._panels.clear();
 
+    if (this.markupview) {
+      this.markupview.destroy();
+    }
+
     if (this.layoutview) {
       this.layoutview.destroy();
     }
 
     if (this.fontinspector) {
       this.fontinspector.destroy();
     }
 
diff --git a/devtools/client/inspector/markup/actions/index.js b/devtools/client/inspector/markup/actions/index.js
new file mode 100644
--- /dev/null
+++ b/devtools/client/inspector/markup/actions/index.js
@@ -0,0 +1,13 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+const { createEnum } = require("devtools/client/shared/enum");
+
+createEnum([
+
+  "UPDATE_MARKUP",
+
+], module.exports);
diff --git a/devtools/client/inspector/markup/actions/moz.build b/devtools/client/inspector/markup/actions/moz.build
new file mode 100644
--- /dev/null
+++ b/devtools/client/inspector/markup/actions/moz.build
@@ -0,0 +1,9 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+DevToolsModules(
+    'index.js',
+)
diff --git a/devtools/client/inspector/markup/components/MarkupApp.js b/devtools/client/inspector/markup/components/MarkupApp.js
new file mode 100644
--- /dev/null
+++ b/devtools/client/inspector/markup/components/MarkupApp.js
@@ -0,0 +1,25 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+const { PureComponent } = require("devtools/client/shared/vendor/react");
+const dom = require("devtools/client/shared/vendor/react-dom-factories");
+const { connect } = require("devtools/client/shared/vendor/react-redux");
+
+class MarkupApp extends PureComponent {
+  static get propTypes() {
+    return {};
+  }
+
+  render() {
+    return dom.div(
+      {
+        id: "markup-container",
+      }
+    );
+  }
+}
+
+module.exports = connect(state => state)(MarkupApp);
diff --git a/devtools/client/inspector/markup/components/moz.build b/devtools/client/inspector/markup/components/moz.build
new file mode 100644
--- /dev/null
+++ b/devtools/client/inspector/markup/components/moz.build
@@ -0,0 +1,9 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+DevToolsModules(
+    'MarkupApp.js',
+)
diff --git a/devtools/client/inspector/markup/moz.build b/devtools/client/inspector/markup/moz.build
--- a/devtools/client/inspector/markup/moz.build
+++ b/devtools/client/inspector/markup/moz.build
@@ -1,16 +1,20 @@
 # -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
 # vim: set filetype=python:
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 DIRS += [
+    'actions',
+    'components',
+    'reducers',
     'views',
 ]
 
 DevToolsModules(
     'markup.js',
+    'new-markup.js',
     'utils.js',
 )
 
 BROWSER_CHROME_MANIFESTS += ['test/browser.ini']
diff --git a/devtools/client/inspector/markup/new-markup.js b/devtools/client/inspector/markup/new-markup.js
new file mode 100644
--- /dev/null
+++ b/devtools/client/inspector/markup/new-markup.js
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+const { createFactory, createElement } = require("devtools/client/shared/vendor/react");
+const { Provider } = require("devtools/client/shared/vendor/react-redux");
+
+const MarkupApp = createFactory(require("./components/MarkupApp"));
+
+const { LocalizationHelper } = require("devtools/shared/l10n");
+const INSPECTOR_L10N =
+  new LocalizationHelper("devtools/client/locales/inspector.properties");
+
+class MarkupView {
+  constructor(inspector, window) {
+    this.document = window.document;
+    this.inspector = inspector;
+    this.store = inspector.store;
+
+    this.init();
+  }
+
+  init() {
+    if (!this.inspector) {
+      return;
+    }
+
+    let markupApp = MarkupApp({});
+
+    let provider = createElement(Provider, {
+      id: "markupview",
+      key: "markupview",
+      store: this.store,
+      title: INSPECTOR_L10N.getStr("inspector.sidebar.ruleViewTitle")
+    }, markupApp);
+
+    // Expose the provider to let inspector.js use it in setupSidebar.
+    this.provider = provider;
+  }
+
+  destroy() {
+    this.document = null;
+    this.inspector = null;
+    this.store = null;
+  }
+}
+
+module.exports = MarkupView;
diff --git a/devtools/client/inspector/markup/reducers/index.js b/devtools/client/inspector/markup/reducers/index.js
new file mode 100644
--- /dev/null
+++ b/devtools/client/inspector/markup/reducers/index.js
@@ -0,0 +1,7 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+exports.markup = require("./markup");
diff --git a/devtools/client/inspector/markup/reducers/markup.js b/devtools/client/inspector/markup/reducers/markup.js
new file mode 100644
--- /dev/null
+++ b/devtools/client/inspector/markup/reducers/markup.js
@@ -0,0 +1,19 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+const INITIAL_MARKUP = [];
+
+let reducers = {
+
+};
+
+module.exports = function (markup = INITIAL_MARKUP, action) {
+  let reducer = reducers[action.type];
+  if (!reducer) {
+    return markup;
+  }
+  return reducer(markup, action);
+};
diff --git a/devtools/client/inspector/markup/reducers/moz.build b/devtools/client/inspector/markup/reducers/moz.build
new file mode 100644
--- /dev/null
+++ b/devtools/client/inspector/markup/reducers/moz.build
@@ -0,0 +1,10 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+DevToolsModules(
+    'index.js',
+    'markup.js',
+)
diff --git a/devtools/client/preferences/devtools.js b/devtools/client/preferences/devtools.js
--- a/devtools/client/preferences/devtools.js
+++ b/devtools/client/preferences/devtools.js
@@ -69,16 +69,18 @@ pref("devtools.inspector.shapesHighlight
 // Enable the Changes View
 pref("devtools.changesview.enabled", false);
 // Enable the Events View
 pref("devtools.eventsview.enabled", false);
 // Enable the Flexbox Inspector panel
 pref("devtools.flexboxinspector.enabled", false);
 // Enable the new Animation Inspector
 pref("devtools.new-animationinspector.enabled", false);
+// Enable the new Markup view
+pref("devtools.markup.new-frontend-enabled", false);
 
 // Grid highlighter preferences
 pref("devtools.gridinspector.gridOutlineMaxColumns", 50);
 pref("devtools.gridinspector.gridOutlineMaxRows", 50);
 pref("devtools.gridinspector.showGridAreas", false);
 pref("devtools.gridinspector.showGridLineNumbers", false);
 pref("devtools.gridinspector.showInfiniteLines", false);
 
@@ -91,17 +93,16 @@ pref("devtools.layout.flexbox.opened", t
 // Whether or not the grid inspector panel is opened in the layout view
 pref("devtools.layout.grid.opened", true);
 
 // By how many times eyedropper will magnify pixels
 pref("devtools.eyedropper.zoom", 6);
 
 // Enable to collapse attributes that are too long.
 pref("devtools.markup.collapseAttributes", true);
-
 // Length to collapse attributes
 pref("devtools.markup.collapseAttributeLength", 120);
 
 // DevTools default color unit
 pref("devtools.defaultColorUnit", "authored");
 
 // Enable the Memory tools
 pref("devtools.memory.enabled", true);
diff --git a/devtools/client/shared/browser-loader.js b/devtools/client/shared/browser-loader.js
--- a/devtools/client/shared/browser-loader.js
+++ b/devtools/client/shared/browser-loader.js
@@ -14,16 +14,17 @@ const BROWSER_BASED_DIRS = [
   "resource://devtools/client/inspector/boxmodel",
   "resource://devtools/client/inspector/changes",
   "resource://devtools/client/inspector/computed",
   "resource://devtools/client/inspector/events",
   "resource://devtools/client/inspector/flexbox",
   "resource://devtools/client/inspector/fonts",
   "resource://devtools/client/inspector/grids",
   "resource://devtools/client/inspector/layout",
+  "resource://devtools/client/inspector/markup",
   "resource://devtools/client/jsonview",
   "resource://devtools/client/shared/source-map",
   "resource://devtools/client/shared/redux",
   "resource://devtools/client/shared/vendor",
 ];
 
 const COMMON_LIBRARY_DIRS = [
   "resource://devtools/client/shared/vendor",
